### PR TITLE
Feat/#a4k5en #a6k6ev

### PR DESCRIPTION
**Pull request by asianswagstorm**

### CODE REVIEW

**Comment by dannylekim** [comment]
Task linked: [CU-a4k5en Profile Page [3 POINTS]](https://app.clickup.com/t/a4k5en)

**Comment by dannylekim** [comment]
Task linked: [CU-a6k6ev Integration](https://app.clickup.com/t/a6k6ev)

File: src/components/BillFilter/__tests__/BillFilter.test.jsx [hunk 1]
```diff
@@ -3,132 +3,283 @@ import BillFilter from "../BillFilter";
...
-          dateOpened: true,
-        }
-      });
+      wrapper = shallow(
+        <BillFilter
+            activeTab = "allBills"
+            applyFiltering = {mockApplyFiltering}
+            currentSorting = "Newest"
+            billStatusFilter={{
+              resolved: false,
```
Comment (dannylekim) [inline src/components/BillFilter/__tests__/BillFilter.test.jsx]: I see this object a lot, you could define it out as a constant and reuse

File: src/components/SearchBar/SearchBar.jsx [hunk 2]
```diff
@@ -92,8 +92,13 @@ class SearchBar extends Component {
     if (!!dateFilters.startDate.value) {
       params = `start=${this.state.dateFilters.startDate.value}`; 
     }
+
+    if (!!dateFilters.startDate.value && !!dateFilters.endDate.value) {
```
Comment (dannylekim) [inline src/components/SearchBar/SearchBar.jsx]: what's the point of using double bang here? 

it's not different than just using 

dateFilters.startDate.value && dateFilters.endDate.value

File: src/components/Sidebar/Sidebar.jsx [hunk 3]
```diff
@@ -34,8 +34,12 @@ class Sidebar extends React.Component {
     this.handleLogoutClick = this.handleLogoutClick.bind(this);
   }
 
+  componentDidMount(){
+    this.state.activeState.dashboard === true && this.props.history.push(`/dashboard`);
```
Comment (dannylekim) [inline src/components/Sidebar/Sidebar.jsx]: is there a reason we doing === true

File: src/components/Sidebar/Sidebar.jsx [hunk 4]
```diff
@@ -34,8 +34,12 @@ class Sidebar extends React.Component {
...
+  }
+
   handleClick(link) {
-    if (this.state.activeState[link]) {
+    if (this.state.activeState[link] === false) {
```
Comment (dannylekim) [inline src/components/Sidebar/Sidebar.jsx:42]: similarly, is there a reason for === false? Is this object going to be anything else but boolean

File: src/views/App/App.jsx [hunk 5]
```diff
@@ -15,16 +15,17 @@ import "./styles.scss";
...
+export default ({hasUser}) => {
+  
   return (
     <div className='App'>
       <BillSnapBackground showWave />
       <BrowserRouter>
         <Switch>
           <React.Fragment>
             <div className='App__container'>
+              {hasUser === true && <Sidebar />}
```
Comment (dannylekim) [inline src/views/App/App.jsx]: hasUser is explicit enough to be a boolean value

File: src/views/Dashboard/Dashboard.jsx [hunk 6]
```diff
@@ -33,7 +33,7 @@ export default class Dashboard extends Component {
 
     return (
       <>
-        {localStorage.getItem("billSnap_token") ? (
+        {this.props.hasUser === true ? (
```
Comment (dannylekim) [inline src/views/Dashboard/Dashboard.jsx]: here too

File: src/views/LandingPage/LandingPage.jsx [hunk 7]
```diff
@@ -10,12 +10,12 @@ import "./styles.scss";
...
 
 export default (props) => {
-  const { history, toggleFormType, formType } = props;
+  const { history, toggleFormType, formType, hasUser } = props;
 
   return (
     <div className="page__landing">
       <Navbar />
-      {!localStorage.getItem("billSnap_token") ? (
+      {hasUser === false ? (
```
Comment (dannylekim) [inline src/views/LandingPage/LandingPage.jsx]: here too

File: src/views/LandingPage/index.js [hunk 8]
```diff
@@ -4,6 +4,7 @@ import {toggleLandingPageForm} from "../../redux/actions/applicationActions";
 
 const mapStateToProps = (state) => ({
   formType: state.application.showRegisterFirst ? "register" : "login",
+  hasUser: (state.users && Object.keys(state.users.userInfo).length > 0) ? true : false,
```
Comment (dannylekim) [inline src/views/LandingPage/index.js]: you definitely don't need to do a tertiary. just do (state.users && Object.keys(state.users.userInfo).length > 0)

**Comment by dannylekim** [comment]
also tiny code smell

File: src/components/Sidebar/Sidebar.jsx (hunk 3 above)
Comment (asianswagstorm) [reply src/components/Sidebar/Sidebar.jsx]: because if I click on profile, this.state.activeState.profile is true. When I refresh this.state.activeState.dashboard is true. but I am still on the Profile page. It has to show the correct page based on the active sidebar state.

File: src/components/Sidebar/Sidebar.jsx (hunk 4 above)
Comment (asianswagstorm) [reply src/components/Sidebar/Sidebar.jsx:42]: because this.state.activeState[link] alone just means if it exists. so it doesn't enter the if statement.

Comment (asianswagstorm) [reply src/components/Sidebar/Sidebar.jsx:42]: I added === false so that if I click on another nav it will redirect

File: src/components/BillFilter/BillFilter.jsx [hunk 9]
```diff
@@ -88,15 +88,15 @@ const BillFilter = ({
...
-          <Badge
-            key={key}
-            className={badgeObject.title === currentSorting && 'active'}
-            outline
-            pill
-            id={badgeObject.id}
-            onClick={badgeObject.onClick}>
-            {badgeObject.title}
-          </Badge>
+          <span key={key}  className={badgeObject.title === currentSorting && 'active'}>
```
Comment (alphakennyn) [inline src/components/BillFilter/BillFilter.jsx]: why did you add a span here?

File: src/components/BillFilter/__tests__/BillFilter.test.jsx [hunk 10]
```diff
@@ -1,134 +1,247 @@
 import React from "react";
 import BillFilter from "../BillFilter";
-import {shallow} from "enzyme";
+import { shallow } from "enzyme";
```
Comment (alphakennyn) [inline src/components/BillFilter/__tests__/BillFilter.test.jsx:3]: Did you pull from release branch?

File: src/components/BillFilter/__tests__/BillFilter.test.jsx [hunk 11]
```diff
@@ -1,134 +1,247 @@
...
-          }
-        })
-        matches(wrapper);
+    });
+  });
+
+  describe("component ", () => {
+    it("should call onClicks of badge clicks", () => {
+      wrapper = shallow(
+        <BillFilter
```
Comment (alphakennyn) [inline src/components/BillFilter/__tests__/BillFilter.test.jsx:79]: Can you put this definition in a `beforeEach`

File: src/components/BillFilter/__tests__/BillFilter.test.jsx [hunk 12]
```diff
@@ -1,134 +1,247 @@
...
+        />
+      );
 
       expect(wrapper.find("FormCheckbox").at(1).render().text()).toBe(
         "End date: 2020-01-01"
       );
     });
+
+    it("should trigger mockDateCheckboxHandler and handleDateSelection", () => {
+      wrapper = shallow(
```
Comment (alphakennyn) [inline src/components/BillFilter/__tests__/BillFilter.test.jsx]: i hope you can see how unreadable these tests are. Please assign the wrapper in a `beforeEach` and simply set the props (`setProps`) in your tests

File: src/components/BillSummary/BillSummary.jsx [hunk 13]
```diff
@@ -8,7 +8,7 @@ class BillSummary extends Component {
 
     return (
       <div className='bill__summary'>
-        {activeBill.id ? (
+        {activeBill && activeBill.id ? (
```
Comment (alphakennyn) [inline src/components/BillSummary/BillSummary.jsx]: remove this. `activeBill` will always be an object. 

`{} == true`

File: src/components/SearchBar/SearchBar.jsx (hunk 2 above)
Comment (alphakennyn) [reply src/components/SearchBar/SearchBar.jsx]: this is just to ensure boolean value. Though this check is a redundant since REST request will still work with a prepended `&`

File: src/components/Sidebar/Sidebar.jsx (hunk 3 above)
Comment (alphakennyn) [reply src/components/Sidebar/Sidebar.jsx]: This did not answer @dannylekim question. He's just asking why you write `true`

just this is ok `this.state.activeState.dashboard && this.props.history.push(`/dashboard`);`

File: src/components/Sidebar/Sidebar.jsx (hunk 4 above)
Comment (alphakennyn) [reply src/components/Sidebar/Sidebar.jsx:42]: `=== false` is definitely redundant here.

File: src/components/Sidebar/Sidebar.test.jsx [hunk 14]
```diff
@@ -43,10 +34,7 @@ describe("Sidebar", () => {
   describe("render", () => {
     describe("snapshots 📸", () => {
       it("Sidebar should match snap shot", () => {
-        matches(<Sidebar />);
-      });
-      it("Sidebar should match snap shot if hide props is true", () => {
```
Comment (alphakennyn) [inline src/components/Sidebar/Sidebar.test.jsx:48]: ??

File: src/views/BillDisplay/BillDisplay.jsx [hunk 15]
```diff
@@ -20,7 +20,7 @@ class BillDisplay extends Component {
     const BillsList = (billsVar) => {
       return (
         <div className="bill__container">
-          {billsVar.length > 0 ? (
```
Comment (alphakennyn) [inline src/views/BillDisplay/BillDisplay.jsx:23]: why?

this is good

File: src/views/Dashboard/index.js [hunk 16]
```diff
@@ -3,6 +3,7 @@ import {connect} from "react-redux";
 
 const mapStateToProps = (state) => ({
   userInfo: state.users.userInfo,
+  hasUser: Object.keys(state.users.userInfo).length > 0 ? true : false,
```
Comment (alphakennyn) [inline src/views/Dashboard/index.js]:
```suggestion
  hasUser: Object.keys(state.users.userInfo).length > 0,
```

**Review by alphakennyn — CHANGES_REQUESTED** [review]

File: src/components/BillFilter/BillFilter.jsx (hunk 9 above)
Comment (asianswagstorm) [reply src/components/BillFilter/BillFilter.jsx]: Badge doesnt take classname or key prop

File: src/components/BillSummary/BillSummary.jsx (hunk 13 above)
Comment (asianswagstorm) [reply src/components/BillSummary/BillSummary.jsx]: not true, this solves can't read id of undefined. activeBill is undefined sometimes

File: src/components/Sidebar/Sidebar.test.jsx (hunk 14 above)
Comment (asianswagstorm) [reply src/components/Sidebar/Sidebar.test.jsx:48]: hide doesn't exists

File: src/views/BillDisplay/BillDisplay.jsx (hunk 15 above)
Comment (asianswagstorm) [reply src/views/BillDisplay/BillDisplay.jsx:23]: its already handled on line 50   bills.length > 0 ? (
            BillsList(bills)
        ) : (
          <p> No Bills Found</p>
        )}

File: src/components/Sidebar/Sidebar.jsx (hunk 4 above)
Comment (asianswagstorm) [reply src/components/Sidebar/Sidebar.jsx:42]: Not true, its neccesary. if its not there it wont enter the if block. and if you do !this.state.activeState[link] it doesnt mean its false, it means is not a key of activeState

File: src/components/BillFilter/__tests__/BillFilter.test.jsx (hunk 10 above)
Comment (asianswagstorm) [reply src/components/BillFilter/__tests__/BillFilter.test.jsx:3]: I merged your PR and rebased

File: src/components/SearchBar/SearchBar.jsx (hunk 2 above)
Comment (dannylekim) [reply src/components/SearchBar/SearchBar.jsx]: that doesn't make sense, because `(dateFilters.startDate.value && dateFilters.endDate.value)` ensures a boolean value on its own

File: src/components/Sidebar/Sidebar.jsx (hunk 4 above)
Comment (dannylekim) [reply src/components/Sidebar/Sidebar.jsx:42]: I find that a bit odd 

if it's a boolean value 

`this.state.activeState[link]: false` 

`(!this.state.activeState[link])` would go in, if it's true it wouldn't. The only time that you should need to explicitly write === false is because you're saying that this.state.activeState[link] can be a value other boolean (such as undefined, null, object, etc), is that the case?

File: src/components/BillFilter/BillFilter.jsx (hunk 9 above)
Comment (dannylekim) [reply src/components/BillFilter/BillFilter.jsx]:
```
const Badge = props => {
  let { tag: Tag, className, theme, pill, outline, ...attrs } = props;

  const classes = classNames(
    className,
    "badge",
    theme && !outline && `badge-${theme}`,
    outline && `badge-outline-${theme}`,
    pill && "badge-pill"
  );

  Tag = attrs.href && Tag === "span" ? "a" : Tag;

  return <Tag {...attrs} className={classes} />;
};
```


it looks to me that it does take className and key

File: src/components/BillSummary/BillSummary.jsx (hunk 13 above)
Comment (dannylekim) [reply src/components/BillSummary/BillSummary.jsx]: Good point

File: src/components/Sidebar/Sidebar.jsx (hunk 4 above)
Comment (asianswagstorm) [reply src/components/Sidebar/Sidebar.jsx:42]: This also passes the test if you try to pass in a random string as link in the handle function. Since that random string is not a key in the active state it will not enter the if block.

File: src/components/BillFilter/BillFilter.jsx (hunk 9 above)
Comment (asianswagstorm) [reply src/components/BillFilter/BillFilter.jsx]: you get a console warning if you add classname and key in the badge

File: src/components/Sidebar/Sidebar.jsx (hunk 4 above)
Comment (dannylekim) [reply src/components/Sidebar/Sidebar.jsx:42]: okay which means there's the possibility in which the value can not be a boolean? Only then is it fine to do this equals (but even then to me it's odd that the value could not be boolean value, but i digress)

File: src/components/SearchBar/SearchBar.jsx (hunk 2 above)
Comment (alphakennyn) [reply src/components/SearchBar/SearchBar.jsx]: in that case, we'll need to be consistent everywhere else and remove redundant double bang

File: src/components/Sidebar/Sidebar.jsx (hunk 4 above)
Comment (alphakennyn) [reply src/components/Sidebar/Sidebar.jsx:42]: IMO i think the activeState should be an stored as a string value (kind of like how we do for activeTab). That way, we can more accurately pinpoint the activeState.

Doing it this way where a `boolean` or a random string are potential params is not so intuitive.

File: src/components/BillFilter/__tests__/BillFilter.test.jsx [hunk 17]
```diff
@@ -1,134 +1,198 @@
...
-          }
-        })
-        matches(wrapper);
+      wrapper.find("Badge").at(2).simulate("click");
+      expect(mockFilterToggleChange).toBeCalledWith({
+        statusOpened: true,
+        categoryOpened: false,
+        dateOpened: false,
       });
-      it("BillFilter should match snap shot if status filter is active", () => {
```
Comment (alphakennyn) [inline src/components/BillFilter/__tests__/BillFilter.test.jsx:76]: why do you remove these snaps?

we want to test that the props reflect on the render

File: src/components/Sidebar/Sidebar.test.jsx (hunk 14 above)
Comment (alphakennyn) [reply src/components/Sidebar/Sidebar.test.jsx:48]: My poor `hide` prop

`<Navbar className="billSnap-SideBar" style={ this.props.hide ? { display: 'none' } : {}}>`

Comment (alphakennyn) [reply src/components/Sidebar/Sidebar.test.jsx:48]: this is the first line in SideBar render

File: src/views/App/App.jsx [hunk 18]
```diff
@@ -15,18 +15,18 @@ import "./styles.scss";
...
         <Switch>
           <React.Fragment>
-            <div className='App__container'>
-              <Route path='/' exact component={LandingPage} />
-              <Sidebar hide={ !(!!localStorage.getItem("billSnap_token")) } />
-              <Route path='/dashboard' exact component={Dashboard} />
-              <Route path='/profile' exact component={Profile} />
+            <div className="App__container">
+              <Route path="/" exact component={LandingPage} />
+              {hasUser && <Sidebar />}
```
Comment (alphakennyn) [inline src/views/App/App.jsx]: ok, if you do the user check here; then remove the hide prop from `SideBar`

File: src/views/Dashboard/index.js [hunk 19]
```diff
@@ -3,6 +3,7 @@ import {connect} from "react-redux";
 
 const mapStateToProps = (state) => ({
   userInfo: state.users.userInfo,
+  hasUser: Object.keys(state.users.userInfo).length > 0,
```
Comment (alphakennyn) [inline src/views/Dashboard/index.js]: can you create a `userSelector` for this?

good practice to have: **_If I have to write this more than once, write a function for it_**

File: src/views/LandingPage/index.js [hunk 20]
```diff
@@ -4,6 +4,7 @@ import {toggleLandingPageForm} from "../../redux/actions/applicationActions";
 
 const mapStateToProps = (state) => ({
   formType: state.application.showRegisterFirst ? "register" : "login",
+  hasUser: (state.users && Object.keys(state.users.userInfo).length > 0),
```
Comment (alphakennyn) [inline src/views/LandingPage/index.js]: wow so many different ways to write it!

File: src/views/Profile/Profile.jsx [hunk 21]
```diff
@@ -5,7 +5,7 @@ import { FaPencilAlt, FaPhone, FaEnvelope, FaMapMarkerAlt, FaBirthdayCake } from
 const capilizeFirstLetter = (name) =>
```
Comment (alphakennyn) [inline src/views/Profile/Profile.jsx]: this slipped me before. Can you extract this into a helper function?

File: src/views/Profile/Profile.jsx [hunk 22]
```diff
@@ -15,30 +15,35 @@ export default ({ userInfo }) => {
     location,
     email,
   } = userInfo;
-  const userFullName = `${capilizeFirstLetter(firstName)} ${capilizeFirstLetter(
+  
+  const userFullName = hasUser ? `${capilizeFirstLetter(firstName)} ${capilizeFirstLetter(
```
Comment (alphakennyn) [inline src/views/Profile/Profile.jsx]: you can extract this into a function too

File: src/views/Profile/index.js [hunk 23]
```diff
@@ -3,6 +3,7 @@ import {connect} from "react-redux";
 
 const mapStateToProps = (state) => ({
   userInfo: state.users.userInfo,
+  hasUser: Object.keys(state.users.userInfo).length > 0,
```
Comment (alphakennyn) [inline src/views/Profile/index.js]: same as above

File: src/components/BillFilter/__tests__/BillFilter.test.jsx [hunk 24]
```diff
@@ -1,134 +1,198 @@
...
+      wrapper.setProps({ filterToggles });
+
+      wrapper.find("FormCheckbox").at(0).simulate("change");
+      expect(mockDateCheckboxHandler).toBeCalledWith("startDate");
+      wrapper.find("FormCheckbox").at(1).simulate("change");
+      expect(mockDateCheckboxHandler).toBeCalledWith("endDate");
+      const event = {
+        target: { value: "2020/01/04" },
+      };
+      wrapper.find("FormInput").at(0).simulate("change", event.target.value);
```
Comment (alphakennyn) [inline src/components/BillFilter/__tests__/BillFilter.test.jsx:204]: shouldn't it be:

```suggestion
      wrapper.find("FormInput").at(0).simulate("change", event);
```

?

Comment (alphakennyn) [reply src/components/BillFilter/__tests__/BillFilter.test.jsx:204]: and if it passed like this, then it goes to show the test is not really proving much here...

**Review by alphakennyn — CHANGES_REQUESTED** [review]

File: src/helpers/Functions.js [hunk 25]
```diff
@@ -28,3 +28,13 @@ export const filterDateTime = (timestamp) => {
   }
   return dateTime;
 };
+
+export const capilizeFirstLetter = (name) =>
```
Comment (alphakennyn) [inline src/helpers/Functions.js]: rather than adding these irrelevant functions witht he date time funciton; please rename the file back to `DateTime` and create a new file called `StringParsers` or `NameParsers`

File: src/helpers/__tests__/helpers.test.js [hunk 26]
```diff
@@ -9,8 +9,16 @@ import {
   FaShoppingCart,
   FaUtensils,
 } from "react-icons/fa";
-
+//hasUser, firstName, middleName, lastName
```
Comment (alphakennyn) [inline src/helpers/__tests__/helpers.test.js]: remove pls

File: src/redux/selectors/userSelectors.js [hunk 27]
```diff
@@ -0,0 +1 @@
+export const getHasUser = (state) => state.users && Object.keys(state.users.userInfo).length > 0;
```
Comment (alphakennyn) [inline src/redux/selectors/userSelectors.js]: this is not how you create a selector.

it should at least pass through `reselect`'s `createSelector` function to use its memoized feature

```js
import { createSelector } from 'reselect'

const loggedUser = (state) => state.users; 

export const getHasUser = createSelector([loggedUser], (user) => user && Object.keys(user.userInfo).length > 0); 
```

Comment (alphakennyn) [reply src/redux/selectors/userSelectors.js]: please confirm that this^ works. I just typed it via github

File: src/helpers/Functions.js [hunk 28]
```diff
@@ -28,3 +28,13 @@ export const filterDateTime = (timestamp) => {
...
+  name.charAt(0).toUpperCase() + name.slice(1);
+
+export const userFullName = (hasUser, firstName, middleName, lastName) =>
```
Comment (alphakennyn) [inline src/helpers/Functions.js]: `userFullName` should not care if user exists, it should only use the last 3 params and format the user's name. Do the validation outside of the function

**Review by alphakennyn — CHANGES_REQUESTED** [review]

**Review by alphakennyn — APPROVED** [review]
